s2sa/encoder.lua

```diff
@@ -0,0 +1,64 @@
+local model_utils = require 's2sa.utils.model_utils'
+local table_utils = require 's2sa.utils.table_utils'
+local cuda = require 's2sa.utils.cuda'
+require 's2sa.sequencer'
+
+local Encoder, Sequencer = torch.class('Encoder', 'Sequencer')
+
+function Encoder:__init(args, network)
+  Sequencer.__init(self, 'enc', args, network)
+
+  -- preallocate context vector
+  self.context_proto = torch.zeros(args.max_batch_size, args.max_sent_length, args.rnn_size)
+
+  -- preallocate output gradients
+  self.grad_out_proto = {}
+  for _ = 1, args.num_layers do
+    table.insert(self.grad_out_proto, torch.zeros(args.max_batch_size, args.rnn_size))
+    table.insert(self.grad_out_proto, torch.zeros(args.max_batch_size, args.rnn_size))
+  end
+end
+
+function Encoder:forward(batch)
```

Comment:
I think we could but what would be the benefits in our case? Here `forward` is mostly a convenience method to factorize the input preparation during training, evaluation and translation.